python/cudf/dataframe/series.py

```diff
@@ -300,6 +301,32 @@ def __getitem__(self, arg):
         else:
             raise NotImplementedError(type(arg))
 
+    def __setitem__(self, key, val):
+        def _process_arg(arg):
+            if isinstance(arg, (list, np.ndarray, pd.Series, range, Index,
+                                DeviceNDArray)):
+                if len(arg) == 0:
+                    return Series([])
+                else:
+                    return Series(arg)
+            return arg
+        key, val = map(_process_arg, (key, val))
+
+        # TODO make this more robust
+        if isinstance(key, Number):
+            key = Series([key])
+            val = Series([val])
+        else:
+            if isinstance(val, Number):
+                val = Series([val] * len(key))
+            else:
+                # TODO something more intelligent here
+                assert len(key) == len(val)
+
+        cpp_copying.apply_scatter_array(to_device(val),
+                                        to_device(key),
+                                        self._column)
```

Comment:
We should move this logic down to the column level so that it can be reused by index columns as well.

Reply:
Meaning implement `__setitem__` at the column level and have `cudf.Series` wrap it?

Reply:
It doesn't have to be `__setitem__` since Column objects aren't user facing, but yes, some API to handle the logic at the column level and then call into it from Series.